tests/test_modeling_realm.py

```diff
@@ -0,0 +1,361 @@
+
+    all_model_classes = (
+        (
+            RealmEmbedder,
+            RealmEncoder,
+            # RealmRetriever is excluded from common tests as it is a container model
+            # consisting of two RealmEmbedders & simple inner product calculation.
+            # RealmRetriever
+        )
+        if is_torch_available()
+        else ()
+    )
+    all_generative_model_classes = ()
+
+    # disable these tests because there is no base_model in Realm
+    test_save_load_fast_init_from_base = False
+    test_save_load_fast_init_to_base = False
+
+    def setUp(self):
+        self.test_pruning = False
+        self.model_tester = RealmModelTester(self)
+        self.config_tester = ConfigTester(self, config_class=RealmConfig)
+
+    def test_config(self):
+        self.config_tester.run_common_tests()
+
+    def test_embedder(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_embedder(*config_and_inputs)
+
+    def test_encoder(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_encoder(*config_and_inputs)
+
+    def test_model_various_embeddings(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        for type in ["absolute", "relative_key", "relative_key_query"]:
+            config_and_inputs[0].position_embedding_type = type
+            self.model_tester.create_and_check_embedder(*config_and_inputs)
+            self.model_tester.create_and_check_encoder(*config_and_inputs)
+
+    def test_retriever(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_retriever(*config_and_inputs)
+
+    def test_training(self):
+        if not self.model_tester.is_training:
+            return
+
+        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+        config.return_dict = True
+
+        for model_class in [RealmEncoder]:
+            model = model_class(config)
+            model.to(torch_device)
+            model.train()
+            inputs = self._prepare_for_class(inputs_dict, model_class, return_labels=True)
+            loss = model(**inputs).loss
+            loss.backward()
+
+    @slow
+    def test_embedder_from_pretrained(self):
+        model = RealmEmbedder.from_pretrained("qqaatw/realm-cc-news-pretrained-embedder")
+        self.assertIsNotNone(model)
+
+    @slow
+    def test_encoder_from_pretrained(self):
+        model = RealmEncoder.from_pretrained("qqaatw/realm-cc-news-pretrained-bert")
+        self.assertIsNotNone(model)
+
+    @slow
+    def test_retriever_from_pretrained(self):
+        model = RealmRetriever.from_pretrained("qqaatw/realm-cc-news-pretrained-retriever")
+        self.assertIsNotNone(model)
+
+
+@require_torch
+class RealmModelIntegrationTest(unittest.TestCase):
```

Comment:
Do you think it could be possible to actually include a "real" end-to-end integration test using Wikipedia as the knowledge corpus? @lhoestq - can we maybe add a REALM index to `datasets`?

Reply:
It's somewhat complex to achieve this for the following reasons:

1. REALM is solely a pre-training method which doesn't contain fine-tuning and predicting methods in its github codebase. The fine-tuning method and predictor are resided in *ORQA* ([paper](https://arxiv.org/abs/1906.00300) and [codebase](https://github.com/google-research/language/tree/master/language/orqa)). Thus, we need to add another model *ORQA* to achieve this.
2. REALM's knowledge retrieving method is somehow like Deep Q Learning does, called `Asynchronous Maximum Inner Product Search (MIPS)`. In other words, it first retrieves top-K probability documents from the corpus to calculate marginal loss; then, after these retrieved documents have slightly staled, a new retrieval starts again. Therefore, I believe these steps need files of script to complete it, so we might not be able to do this in the test.
3. An another way that prevents the complex process mentioned above is directly use pre-trained evidence-block (retrieved documents represented as a (num_block_records, retriever_proj_size) shape tensor), but the model function of this is also placed in *ORQA*.

Therefore, I think that we eventually need to add `ORQA` in order to achieve real end-to-end prediction, and this might be difficult to be written in the integration test.

Reply:
Hello @patrickvonplaten, I've written a PyTorch reimpl. of real end-to-end question answering script using ORQA models with pretrained REALM weights: https://github.com/qqaatw/pytorch-realm-orqa.

Do you think this way meets your hope? Also, should we add an ORQA model in transformers rather than directly combine ORQA with REALM model?  I personally perceive adding ORQA as an independent model is a better choice.

Reply:
The codebase looks awesome - it would be great if we could add a short version of your code as an integration test.  It looks it's sufficient to just use `RealmSearcher` and `RealmReader` - so let's just add those to the integration test so that we have some verification that the model works correctly? Do you think it's possible to convert the tf checkpoints to PT and upload them to the hub so that it's sufficient to just call:

```python
1. searcher = RealmSearcher.from_pretrained(pretrained_path, args.block_records_path)
2. reader = RealmReader.from_pretrained(pretrained_path)
3. tokenizer = RealmTokenizer.from_pretrained("qqaatw/realm-cc-news-pretrained-embedder", do_lower_case=True)
```

? I can help you with uploading the weights in PyTorch

@lhoestq - do you think we could add the block records path: https://github.com/qqaatw/pytorch-realm-orqa/blob/481216728f3cb0bc0d920b4152d9327db8ec2b58/predictor.py#L60 to `datasets`?

Reply:
Ah I see `RealmSearcher` and `RealmReader` are not added to the PR yet - do you think they shouldn't be added because their logic was introduced by ORQA?

Reply:
Answered in more detail further below

Reply:
Sure we can add those model into integration test to test end-to-end predictions. I can convert and upload it to the hub as long as the concerns mentioned below are resolved.